### DISCUSSION TITLE
Variable and device states

### DISCUSSION THREAD

**Discussion by clavay** [pyscada/PyScada#187] [Ideas]
I want to implement a way to define, use and access variable and device error/state.

This could lead to:

- know if there is an error while:
  - connecting to the device,
  - reading a value,
  - writing a value,
- display the state:
  - last state and as time series in the user interface for defined user groups,
  - should we use:
    - the `Variable` model with `RecordedData` or another database model,
    - the `VariableProperty` model (need to add a way to store historical data),
- integrate status change in event management and take decision for some state like:
  - stop reading/writing after some retries,
  - delay reading/writing,
  - send message in user interface/email,
- it should be easy to manage from the admin and user interfaces (auto created, quick state access)

@trombastic: What do you think ?

**Comment by ShreyanshMalewar** [comment]


**Reply by clavay** [reply]
Hello,
Can you give me more information about this and how to implement it ?

**Reply by clavay** [reply]
I think react is not needed for this, JS is enough for transmitting the data to the client.

The discussion here is more about the model and the logic on the server.

**Comment by trombastic** [comment]
my opinion is to integrate the History part into the event management because the it is separate from the time series DB and these events should occur not so often, and the current state either as a field into the Variable Model (my preference) or as a  VariableProperty. The event generation then could be done by a on change signal from the Variable Model.

**Reply by clavay** [reply]
Should we store the Device and Variable states in 2 ways ?

1. Last state in a `state` (string?) field in de `Device` and `Variable` models.
2. State history: 
   - How do you want to manage the history for variable or device states in the event model ?
   - Managing the variable or device state as a (new) variable allows to integrate it easily in an event as any variable.
